src/solution.jl

```diff
@@ -16,113 +13,74 @@ scalar_fn(x::SparseTape) = scalar_fn(to_vaf(x))
+    solve!(
+        problem::Problem,
+        optimizer_factory;
         silent_solver = false,
     )
 
-Solves the problem, populating `problem.optval` with the optimal value,
-as well as the values of the variables (accessed by [`evaluate`](@ref))
-and constraint duals (accessed by `cons.dual`), where applicable.
+Solves the problem, populating `problem.optval` with the optimal value, as well
+as the values of the variables (accessed by [`evaluate`](@ref)) and constraint
+duals (accessed by `cons.dual`), where applicable.
+
 Optional keyword arguments:
-* `silent_solver`: whether the solver should be silent (and not emit output or logs) during the solution process.
+
+ * `silent_solver`: whether the solver should be silent (and not emit output or
+   logs) during the solution process.
 """
 function solve!(p::Problem, optimizer_factory; silent_solver = false)
-    context = Context(p, optimizer_factory)
-    model = context.model
-
-    if silent_solver
-        MOI.set(model, MOI.Silent(), true)
-    end
-
-    # check DCPness
     if problem_vexity(p) in (ConcaveVexity(), NotDcp())
         throw(DCPViolationError())
     end
-
+    context = Context(p, optimizer_factory)
+    if silent_solver
+        MOI.set(context.model, MOI.Silent(), true)
+    end
     if context.detected_infeasible_during_formulation[]
         p.status = MOI.INFEASIBLE
     else
-        MOI.optimize!(model)
-        p.status = MOI.get(model, MOI.TerminationStatus())
+        MOI.optimize!(context.model)
+        p.status = MOI.get(context.model, MOI.TerminationStatus())
     end
-
-    p.model = model
-
+    p.model = context.model
     if p.status != MOI.OPTIMAL
         @warn "Problem wasn't solved optimally" status = p.status
     end
-
-    dual_status = MOI.get(model, MOI.DualStatus())
-    primal_status = MOI.get(model, MOI.PrimalStatus())
-
-    var_to_indices = context.var_id_to_moi_indices
-    id_to_variables = context.id_to_variables
-
-    if primal_status != MOI.NO_SOLUTION
-        for (id, var_indices) in var_to_indices
-            var = id_to_variables[id]
-            vexity(var) == ConstVexity() && continue
-            if var_indices isa Tuple
-                vectorized_value_re =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices[1])
-                vectorized_value_im =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices[2])
-                set_value!(
-                    var,
-                    unpackvec(vectorized_value_re, size(var), false) +
-                    im * unpackvec(vectorized_value_im, size(var), false),
-                )
-            else
-                vectorized_value =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices)
-                set_value!(
-                    var,
-                    unpackvec(
-                        vectorized_value,
-                        size(var),
-                        iscomplex(sign(var)),
```

Comment:
Okay here's a tricky one. This branch occurs if `var_indices` is not a tuple, which means it must be a `Vector{MOI.VariableIndex}`. If so, in what situation is `sign(var)` complex? Because the `add_variables!` method adds a tuple if the sign is complex.

Is the `AbstractVariable` API intended to be extensible by third-party packages? Is it used? Are there any examples that we should be checking for breakage?

Reply:
Uhh, I think if it wasn’t complex we’d get an error anyway inside unpackvec since the sizes wouldn’t work. So seems safe to say it must be real.

we do support custom AbstractVariables, but afaict no one has ever done that. I do have tests to ensure the api is working and I think we do solve a problem there. Maybe only in the real case though

Reply:
I’ll try to take a closer look at these tomorrow

Reply:
Ok, I had another look, and I think the diff here is fine if the tests pass, but I don't 100% remember why it was written like this. We do test [3 kinds of custom AbstractVariables](https://github.com/jump-dev/Convex.jl/blob/946d18d131a3404c984725fbd05cae1bf38d0bc4/test/test_utilities.jl#L744-L917) so I think that should cover us pretty well.

BTW I had bigger plans for the AbstractVariable stuff to build higher level functionality / DSLs on top of Convex such as for quantum information theory, so that's where that was going, but I think the real useful feature is adding constraints to a `Variable` itself (so you have a function like `make_probability_vector(d)` that doesn't need the model as an input and the result of which can be reused between models) and the full AbstractVariable thing isn't necessarily much more useful than that.

Reply:
Yes, I've been mulling over https://github.com/jump-dev/Convex.jl/issues/310